File: api/live.py
"""
Vercel serverless function: /api/live
Returns LIVE game scores + player box-score stats from ESPN.

Endpoints:
  GET /api/live?sport=basketball&league=nba
      → all live/recent games with scores and player stats

  GET /api/live?sport=basketball&league=nba&event_id=401585857
      → single game full box score

  GET /api/live?sport=soccer&league=eng.1
  GET /api/live?sport=soccer&league=esp.1
  GET /api/live?sport=football&league=nfl
  GET /api/live?sport=baseball&league=mlb
  GET /api/live?sport=hockey&league=nhl

Query params:
  sport      basketball | football | baseball | hockey | soccer
  league     nba | nfl | mlb | nhl | eng.1 | esp.1 | ita.1 | ger.1 | mex.1 | usa.1 | uefa.champions
  event_id   ESPN event ID (optional — returns single game)
  top        max players per team to return (default 8)
"""

from http.server import BaseHTTPRequestHandler
from urllib.parse import urlparse, parse_qs
import json, urllib.request

ESPN_SITE = "https://site.api.espn.com/apis/site/v2/sports"

# ─── Keys ESPN uses in box-score stats per sport ─────────────────────────────
STAT_KEYS = {
    "basketball": {
        "order": ["MIN","PTS","REB","AST","3PM","FG","FT","STL","BLK","TO","PF"],
        "labels": {
            "MIN":"MIN","PTS":"PTS","REB":"REB","AST":"AST","3PM":"3PM",
            "FG":"TC","FT":"TL","STL":"ROB","BLK":"TAP","TO":"PÉR","PF":"FALTAS",
        },
    },
    "football": {
        "order": ["PYDS","PTD","INT","RUYDS","RUTD","REC","REYDS","RETD","SACKS"],
        "labels": {
            "PYDS":"Yds pase","PTD":"TD pase","INT":"INT",
            "RUYDS":"Yds corrida","RUTD":"TD corrida",
            "REC":"Rec.","REYDS":"Yds rec.","RETD":"TD rec.","SACKS":"Sacks",
        },
    },
    "baseball": {
        "order": ["AB","R","H","HR","RBI","BB","SO","AVG","ERA","IP","K","BB2"],
        "labels": {
            "AB":"VB","R":"Carreras","H":"Hits","HR":"HR","RBI":"RBI",
            "BB":"BB","SO":"K","AVG":"AVG","ERA":"ERA","IP":"IP","K":"K","BB2":"BB",
        },
    },
    "hockey": {
        "order": ["G","A","PTS","+/-","PIM","SOG","TOI","SV","GA","SV%"],
        "labels": {
            "G":"Goles","A":"Asis.","PTS":"PTS","+/-":"+/-","PIM":"PIM",
            "SOG":"Tiros","TOI":"Tiempo","SV":"Paradas","GA":"GA","SV%":"SV%",
        },
    },
    "soccer": {
        "order": ["G","A","SH","SHG","FC","YC","RC","OFF","MIN"],
        "labels": {
            "G":"Goles","A":"Asis.","SH":"Tiros","SHG":"Al arco",
            "FC":"Faltas","YC":"Amarillas","RC":"Rojas","OFF":"Fuera juego","MIN":"MIN",
        },
    },
}


def _get(url, params=None):
    if params:
        qs = "&".join(f"{k}={v}" for k, v in params.items())
        url = f"{url}?{qs}"
    req = urllib.request.Request(url, headers={"User-Agent": "ScoutAI/1.0"})
    with urllib.request.urlopen(req, timeout=12) as r:
        return json.loads(r.read())


def parse_status(comp):
    """Returns (status_str, clock_str, is_live, is_final)."""
    st = comp.get("status", {})
    stype = st.get("type", {})
    name  = stype.get("name", "")
    desc  = stype.get("description", "")
    clock = st.get("displayClock", "")
    period = st.get("period", 0)

    is_live  = name in ("STATUS_IN_PROGRESS",)
    is_final = name in ("STATUS_FINAL", "STATUS_FULL_TIME", "STATUS_END_PERIOD")

    if is_live:
        status = f"EN VIVO · P{period} {clock}"
    elif is_final:
        status = "FINAL"
    else:
        status = desc or name

    return status, clock, is_live, is_final


def parse_competitor(comp_data):
    """Extract team name, score, logo from a competitor object."""
    team = comp_data.get("team", {})
    return {
        "id":     team.get("id", ""),
        "name":   team.get("displayName") or team.get("name", ""),
        "abbr":   team.get("abbreviation", ""),
        "logo":   team.get("logo", ""),
        "score":  comp_data.get("score", "—"),
        "home":   comp_data.get("homeAway", "") == "home",
    }


def parse_box_score(event_id, sport, league, top):
    """Fetch game summary and extract box score players."""
    data = _get(f"{ESPN_SITE}/{sport}/{league}/summary", {"event": event_id})

    sk = STAT_KEYS.get(sport, STAT_KEYS["basketball"])
    want_order  = sk["order"]
    want_labels = sk["labels"]

    teams_out = []
    for team_block in data.get("boxscore", {}).get("players", []):
        team_info = team_block.get("team", {})
        team_name = team_info.get("displayName") or team_info.get("name", "")
        team_logo = team_info.get("logo", "")

        players_out = []
        for stat_block in team_block.get("statistics", []):
            stat_names = stat_block.get("names", [])   # column headers
            athletes   = stat_block.get("athletes", [])

            for ath in athletes[:top]:
                ath_info = ath.get("athlete", {})
                raw_stats = ath.get("stats", [])

                stats_dict = {}
                for i, val in enumerate(raw_stats):
                    if i < len(stat_names):
                        col = stat_names[i]
                        if col in want_labels:
                            stats_dict[col] = val

                # Reorder
                ordered = {k: stats_dict[k] for k in want_order if k in stats_dict}

                pos_d = ath_info.get("position", {})
                players_out.append({
                    "id":       str(ath_info.get("id", "")),
                    "name":     ath_info.get("displayName") or ath_info.get("shortName", ""),
                    "jersey":   ath_info.get("jersey", ""),
                    "position": pos_d.get("abbreviation", "") if isinstance(pos_d, dict) else "",
                    "headshot": (ath_info.get("headshot") or {}).get("href", ""),
                    "starter":  ath.get("starter", False),
                    "stats":    ordered,
                })

        teams_out.append({
            "name":    team_name,
            "logo":    team_logo,
            "players": players_out,
        })

    return teams_out, want_labels


def fetch_scoreboard(sport, league, top):
    """Return all live/recent games with basic info + box scores."""
    data = _get(f"{ESPN_SITE}/{sport}/{league}/scoreboard")
    sk = STAT_KEYS.get(sport, STAT_KEYS["basketball"])

    games_out = []
    for ev in data.get("events", []):
        comps = ev.get("competitions", [{}])
        comp  = comps[0] if comps else {}

        competitors = comp.get("competitors", [])
        teams = [parse_competitor(c) for c in competitors]

        status, clock, is_live, is_final = parse_status(comp)

        event_id = str(ev.get("id", ""))
        venue_d  = comp.get("venue", {})

        # Fetch box score only for live/final games (saves quota for scheduled)
        players_by_team = []
        if (is_live or is_final) and event_id:
            try:
                players_by_team, _ = parse_box_score(event_id, sport, league, top)
            except Exception:
                players_by_team = []

        games_out.append({
            "event_id":   event_id,
            "name":       ev.get("name", ""),
            "date":       ev.get("date", ""),
            "status":     status,
            "is_live":    is_live,
            "is_final":   is_final,
            "venue":      venue_d.get("fullName", ""),
            "teams":      teams,
            "box_score":  players_by_team,
        })

    return {
        "sport":       sport,
        "league":      league,
        "stat_labels": sk["labels"],
        "games":       games_out,
    }


class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        parsed = urlparse(self.path)
        qs     = parse_qs(parsed.query)

        def p(k, default=""):
            v = qs.get(k, [default])
            return v[0] if v else default

        sport    = p("sport", "basketball")
        league   = p("league", "nba")
        event_id = p("event_id", "")
        top      = min(int(p("top", "8")), 20)

        resp_headers = {
            "Content-Type":                "application/json",
            "Access-Control-Allow-Origin": "*",
            # Live data: short cache (30s), stale-while-revalidate 60s
            "Cache-Control":               "s-maxage=30, stale-while-revalidate=60",
        }

        try:
            if event_id:
                # Single game box score
                sk = STAT_KEYS.get(sport, STAT_KEYS["basketball"])
                players_by_team, labels = parse_box_score(event_id, sport, league, top)
                data = {
                    "sport":       sport,
                    "league":      league,
                    "event_id":    event_id,
                    "stat_labels": labels,
                    "box_score":   players_by_team,
                }
            else:
                # Full scoreboard
                data = fetch_scoreboard(sport, league, top)

            body = json.dumps(data, ensure_ascii=False)
            self.send_response(200)
            for k, v in resp_headers.items():
                self.send_header(k, v)
            self.end_headers()
            self.wfile.write(body.encode())

        except Exception as e:
            err = json.dumps({"error": str(e)}, ensure_ascii=False)
            self.send_response(400)
            for k, v in resp_headers.items():
                self.send_header(k, v)
            self.end_headers()
            self.wfile.write(err.encode())

    def do_OPTIONS(self):
        self.send_response(204)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Access-Control-Allow-Methods", "GET, OPTIONS")
        self.send_header("Access-Control-Allow-Headers", "Content-Type")
        self.end_headers()

    def log_message(self, *_):
        pass

File: api/live.js
/**
 * /api/live — Marcador en vivo + box score de jugadores (ESPN)
 *
 * GET /api/live?sport=basketball&league=nba
 *     → todos los partidos (en vivo / recientes) con stats de jugadores
 *
 * GET /api/live?sport=basketball&league=nba&event_id=401585857
 *     → box score completo de un partido específico
 *
 * Ligas disponibles:
 *   basketball: nba
 *   football:   nfl
 *   baseball:   mlb
 *   hockey:     nhl
 *   soccer:     eng.1 | esp.1 | ita.1 | ger.1 | fra.1 | mex.1 | usa.1 | uefa.champions
 *
 * Params:
 *   sport     basketball | football | baseball | hockey | soccer
 *   league    ver arriba
 *   event_id  ESPN event ID (opcional)
 *   top       max jugadores por equipo (default 8, max 20)
 */

const ESPN_SITE = "https://site.api.espn.com/apis/site/v2/sports";

const STAT_KEYS = {
  basketball: {
    order:  ["MIN","PTS","REB","AST","3PM","FG","FT","STL","BLK","TO"],
    labels: { MIN:"MIN",PTS:"PTS",REB:"REB",AST:"AST","3PM":"3PM",FG:"TC",FT:"TL",STL:"ROB",BLK:"TAP",TO:"PÉR" },
  },
  football: {
    order:  ["PYDS","PTD","INT","RUYDS","RUTD","REC","REYDS","RETD","SACKS"],
    labels: { PYDS:"Yds pase",PTD:"TD pase",INT:"INT",RUYDS:"Yds corrida",RUTD:"TD corrida",
              REC:"Rec.",REYDS:"Yds rec.",RETD:"TD rec.",SACKS:"Sacks" },
  },
  baseball: {
    order:  ["AB","R","H","HR","RBI","BB","SO","AVG","ERA","IP","K"],
    labels: { AB:"VB",R:"Carreras",H:"Hits",HR:"HR",RBI:"RBI",BB:"BB",SO:"K",
              AVG:"AVG",ERA:"ERA",IP:"IP",K:"K" },
  },
  hockey: {
    order:  ["G","A","PTS","+/-","PIM","SOG","TOI","SV","GA","SV%"],
    labels: { G:"Goles",A:"Asis.",PTS:"PTS","+/-":"+/-",PIM:"PIM",
              SOG:"Tiros",TOI:"Tiempo",SV:"Paradas",GA:"GA","SV%":"SV%" },
  },
  soccer: {
    order:  ["G","A","SH","SHG","FC","YC","RC","OFF","MIN"],
    labels: { G:"Goles",A:"Asis.",SH:"Tiros",SHG:"Al arco",
              FC:"Faltas",YC:"Amarillas",RC:"Rojas",OFF:"Fuera juego",MIN:"MIN" },
  },
};

// ─── Helpers ─────────────────────────────────────────────────────────────────

async function espnGet(url, params = {}) {
  const qs = new URLSearchParams(params).toString();
  const full = qs ? `${url}?${qs}` : url;
  const res = await fetch(full, { headers: { "User-Agent": "ScoutAI/1.0" } });
  if (!res.ok) throw new Error(`ESPN ${res.status}: ${full}`);
  return res.json();
}

function parseStatus(comp) {
  const st    = comp.status ?? {};
  const stype = st.type ?? {};
  const name  = stype.name ?? "";
  const clock = st.displayClock ?? "";
  const period = st.period ?? 0;

  const isLive  = name === "STATUS_IN_PROGRESS";
  const isFinal = ["STATUS_FINAL","STATUS_FULL_TIME","STATUS_END_PERIOD"].includes(name);

  let status;
  if (isLive)       status = `EN VIVO · P${period} ${clock}`;
  else if (isFinal) status = "FINAL";
  else              status = stype.description ?? name;

  return { status, clock, period, isLive, isFinal };
}

function parseCompetitor(c) {
  const team = c.team ?? {};
  return {
    id:    team.id ?? "",
    name:  team.displayName ?? team.name ?? "",
    abbr:  team.abbreviation ?? "",
    logo:  team.logo ?? "",
    score: c.score ?? "—",
    home:  c.homeAway === "home",
  };
}

async function fetchBoxScore(eventId, sport, league, top) {
  const data = await espnGet(`${ESPN_SITE}/${sport}/${league}/summary`, { event: eventId });
  const sk = STAT_KEYS[sport] ?? STAT_KEYS.basketball;

  const teamsList = [];
  for (const teamBlock of data?.boxscore?.players ?? []) {
    const teamInfo = teamBlock.team ?? {};
    const players  = [];

    for (const statBlock of teamBlock.statistics ?? []) {
      const cols     = statBlock.names ?? [];
      const athletes = statBlock.athletes ?? [];

      for (const ath of athletes.slice(0, top)) {
        const info     = ath.athlete ?? {};
        const rawStats = ath.stats ?? [];

        const statsMap = {};
        rawStats.forEach((val, i) => {
          const col = cols[i];
          if (col && sk.labels[col] !== undefined) statsMap[col] = val;
        });

        // Ordered stats (only keys present)
        const ordered = {};
        for (const k of sk.order) {
          if (statsMap[k] !== undefined) ordered[k] = statsMap[k];
        }

        players.push({
          id:       String(info.id ?? ""),
          name:     info.displayName ?? info.shortName ?? "",
          jersey:   info.jersey ?? "",
          position: info.position?.abbreviation ?? "",
          headshot: info.headshot?.href ?? "",
          starter:  ath.starter ?? false,
          stats:    ordered,
        });
      }
    }

    teamsList.push({
      id:      teamInfo.id ?? "",
      name:    teamInfo.displayName ?? teamInfo.name ?? "",
      logo:    teamInfo.logo ?? "",
      players,
    });
  }

  return { teams: teamsList, labels: sk.labels };
}

async function fetchScoreboard(sport, league, top) {
  const data = await espnGet(`${ESPN_SITE}/${sport}/${league}/scoreboard`);
  const sk   = STAT_KEYS[sport] ?? STAT_KEYS.basketball;
  const games = [];

  for (const ev of data.events ?? []) {
    const comp  = ev.competitions?.[0] ?? {};
    const { status, clock, period, isLive, isFinal } = parseStatus(comp);
    const teams = (comp.competitors ?? []).map(parseCompetitor);
    const eventId = String(ev.id ?? "");

    let boxScore = [];
    if ((isLive || isFinal) && eventId) {
      try {
        const { teams: t } = await fetchBoxScore(eventId, sport, league, top);
        boxScore = t;
      } catch { /* skip */ }
    }

    games.push({
      event_id:  eventId,
      name:      ev.name ?? "",
      date:      ev.date ?? "",
      status,
      clock,
      period,
      is_live:   isLive,
      is_final:  isFinal,
      venue:     comp.venue?.fullName ?? "",
      teams,
      box_score: boxScore,
    });
  }

  return { sport, league, stat_labels: sk.labels, games };
}

// ─── Main handler ─────────────────────────────────────────────────────────────

export default async function handler(req, res) {
  res.setHeader("Access-Control-Allow-Origin", "*");
  res.setHeader("Access-Control-Allow-Methods", "GET, OPTIONS");
  if (req.method === "OPTIONS") return res.status(204).end();

  const { sport = "basketball", league = "nba", event_id = "", top = "8" } = req.query;
  const topN = Math.min(parseInt(top) || 8, 20);

  // Live data: very short cache
  res.setHeader("Cache-Control", "s-maxage=30, stale-while-revalidate=60");

  try {
    let data;
    if (event_id) {
      const sk = STAT_KEYS[sport] ?? STAT_KEYS.basketball;
      const { teams, labels } = await fetchBoxScore(event_id, sport, league, topN);
      data = { sport, league, event_id, stat_labels: labels, box_score: teams };
    } else {
      data = await fetchScoreboard(sport, league, topN);
    }
    return res.status(200).json(data);
  } catch (err) {
    return res.status(400).json({ error: err.message });
  }
}
